### PR TITLE
[Issue 13854][Broker]Fix call sync method in async rest api for internalExpireMessagesForA…

### PR DESCRIPTION
**Pull request by AnonHxy**
Master Issue: <https://github.com/apache/pulsar/issues/13854>

### Motivation

See Issue: <https://github.com/apache/pulsar/issues/13854>

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation
  
- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -3445,6 +3449,92 @@ private void internalExpireMessagesByTimestampForSinglePartition(String subName,
         }
     }
 
+    private CompletableFuture<Void> internalExpireMessagesByTimestampForSinglePartitionAsync(String subName,
```
Comment (liudezhi2098) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3452]: conflict with  #13880

Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3452]: comment in <https://github.com/apache/pulsar/pull/13880/files>

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -1801,124 +1801,128 @@ protected void internalSkipMessages(String subName, int numMessages, boolean aut
...
+                                asyncResponse.resume(Response.noContent().build());
+                                return null;
+                            });
+                        } else {
+                            return internalExpireMessagesForAllSubscriptionsForNonPartitionedTopicAsync(asyncResponse,
+                                expireTimeInSeconds, authoritative);
                         }
-
-                        asyncResponse.resume(Response.noContent().build());
+                    }).exceptionally(ex -> {
```
[the remark this answers is not in the document]
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1856]: This `exceptionally` can handle `getPartitionedTopicMetadataAsync` exceptions  if  `topicName.isPartitioned() ` return  false

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (liudezhi2098) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3452]: Both PR's modify this method，warning: There may be conflict.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -3445,6 +3449,92 @@ private void internalExpireMessagesByTimestampForSinglePartition(String subName,
...
+        if (topicName.isGlobal()) {
+            future = validateGlobalNamespaceOwnershipAsync(namespaceName);
+        } else {
+            future = CompletableFuture.completedFuture(null);
+        }
+
+        return future.thenCompose(__ -> {
+            // If the topic name is a partition name, no need to get partition topic metadata again
+            if (!topicName.isPartitioned()) {
+                getPartitionedTopicMetadataAsync(topicName, authoritative, false)
```
[the remark this answers is not in the document]
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3465]: Yes, there miss return

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 4]
```diff
@@ -3445,6 +3449,92 @@ private void internalExpireMessagesByTimestampForSinglePartition(String subName,
...
+                    .thenCompose(metadata -> {
+                        if (metadata.partitions > 0) {
+                            String msg = "This method should not be called for partitioned topic";
+                            log.error("[{}] {} {} {}", clientAppId(), msg, topicName, subName);
+                            throw new IllegalStateException(msg);
+                        } else {
+                            return doInternalExpireMessagesByTimestampForSinglePartitionAsync(subName,
```
[the remark this answers is not in the document]
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3472]: If `(!topicName.isPartitioned()) && metadata.partitions == 0` , we should also `doInternalExpireMessagesByTimestampForSinglePartitionAsync`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1856]: > This `exceptionally` can handle `getPartitionedTopicMetadataAsync` exceptions if `topicName.isPartitioned() ` return false

Yes, but exceptionally is also needed there for `validateGlobalNamespaceOwnership`

[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1856]: +1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 5]
```diff
@@ -1801,124 +1801,129 @@ protected void internalSkipMessages(String subName, int numMessages, boolean aut
...
+        AsyncResponse asyncResponse,
+        int expireTimeInSeconds,
+        boolean authoritative) {
+        // validate ownership and redirect if current broker is not owner
+        return validateTopicOwnershipAsync(topicName, authoritative)
+            .thenCompose(__ -> validateTopicOperationAsync(topicName, TopicOperation.EXPIRE_MESSAGES))
+            .thenCompose(__ -> getTopicReferenceAsync(topicName))
+            .thenCompose(t -> {
+                PersistentTopic topic = (PersistentTopic) t;
+                final AtomicReference<Throwable> exception = new AtomicReference<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1877]: Why do we need this line? You can get the exception from the `futures` from line 1879 directly.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3453]: Oh, I think we'd better close the new one and apply the fix from the old one.

Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3453]: Ok,  I can rebase the code after https://github.com/apache/pulsar/pull/13880 merging to master to fix the conflict

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: update

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 6]
```diff
@@ -1801,124 +1801,128 @@ protected void internalSkipMessages(String subName, int numMessages, boolean aut
...
+                    futures.add(
+                    internalExpireMessagesByTimestampForSinglePartitionAsync(subName, expireTimeInSeconds,
+                        authoritative)
+                        .exceptionally(ex -> {
+                            exception.set(FutureUtil.unwrapCompletionException(ex));
+                            return null;
+                        })));
 
-        asyncResponse.resume(Response.noContent().build());
+                return FutureUtil.waitForAll(futures).thenRun(() -> {
```
[the remark this answers is not in the document]
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: update

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 5 above)
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: looks the same as https://github.com/apache/pulsar/pull/13905#discussion_r790265218?  I will update there code

**Pull request closed without merging**